### PR TITLE
ROX-25352: log feature flags

### PR DESCRIPTION
**Pull request by stehessel**
### Description

Add logging to our feature flag framework. Upon service start up, a log line such as this one is produced

```pkg/features: 2024/07/24 17:29:52.180003 features.go:82: Info: Feature flags [dev-preview] {"ROX_ACSCS_EMAIL_NOTIFIER": true, "ROX_ACTIVE_VULN_MGMT": false, "ROX_AGGREGATE_DEPLOYMENT_REFERENCE_OPTIMIZATION": false, "ROX_ATTEMPT_MANIFEST_DIGEST": true, "ROX_AUTH_MACHINE_TO_MACHINE": true, "ROX_CAPTURE_INTERMEDIATE_EVENTS": true, "ROX_CLUSTER_AWARE_DEPLOYMENT_CHECK": true, "ROX_COLLECTOR_RUNTIME_CONFIG": false, "ROX_COMPLIANCE_ENHANCEMENTS": true, "ROX_COMPLIANCE_HIERARCHY_CONTROL_DATA": true, "ROX_COMPLIANCE_REMEDIATION": true, "ROX_COMPLIANCE_REPORTING": true, "ROX_DEFAULT_FALSE": false, "ROX_DEFAULT_TRUE": true, "ROX_DELEGATE_WATCHED_IMAGE_REPROCESSING": true, "ROX_DEPLOYMENT_BUILD_OPTIMIZATION": true, "ROX_DEPLOYMENT_ENVVAR_SEARCH": true, "ROX_DEPLOYMENT_SECRET_SEARCH": true, "ROX_DEPLOYMENT_VOLUME_SEARCH": true, "ROX_POLICY_CRITERIA_MODAL": false, "ROX_PREVENT_SENSOR_RESTART_ON_DISCONNECT": true, "ROX_SCANNER_V4": false, "ROX_SCANNER_V4_MULTI_BUNDLE": true, "ROX_SCANNER_V4_SUPPORT": true, "ROX_SECRET_FILE_SEARCH": true, "ROX_SENSOR_RECONCILIATION": true, "ROX_SENSOR_SINGLE_SCAN": true, "ROX_SOURCED_AUTOGENERATED_INTEGRATIONS": false, "ROX_STORE_EVENT_HASHES": true, "ROX_TEST_FEAT": true, "ROX_UNQUALIFIED_SEARCH_REGISTRIES": false, "ROX_VULN_MGMT_2_GA": true, "ROX_VULN_MGMT_ADVANCED_FILTERS": true, "ROX_VULN_MGMT_LEGACY_SNOOZE": false, "ROX_VULN_MGMT_NODE_PLATFORM_CVES": true, "ROX_VULN_MGMT_REPORTING_ENHANCEMENTS": true, "ROX_VULN_MGMT_UNIFIED_CVE_DEFERRAL": true, "ROX_VULN_MGMT_WORKLOAD_CVES": true, "ROX_WORKLOAD_CVES_FIXABILITY_FILTERS": true}```

### User-facing documentation

(*must be* 2 items and both *must be* checked)

- [ ] CHANGELOG is updated
- [x] CHANGELOG update is not needed
- [ ] Documentation PR is created and linked above
- [x] Documentation is not needed

### Testing

- [ ] inspected CI results

#### Automated testing

(*must be* at least 1 item and all items *must be* checked)

- [x] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests
- [ ] contributed **no automated tests**
  

#### How I validated my change

Looked at the logs.

### CODE REVIEW

File: pkg/features/feature.go [hunk 1]
```diff
@@ -38,3 +41,12 @@ func (f *feature) Enabled() bool {
...
 }
+
+func (f *feature) LoggingContext() zapcore.Field {
+	return logging.Any(f.EnvVar(), f)
+}
+
+func (f *feature) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddBool("enabled", f.Enabled())
+	return nil
+}
```
Comment (janisz) [inline pkg/features/feature.go:52]: Why do we need it here? We can do something similar as you did with telemetry

File: pkg/features/feature.go [hunk 2]
```diff
@@ -3,6 +3,9 @@ package features
 import (
 	"os"
 	"strings"
+
+	"github.com/stackrox/rox/pkg/logging"
+	"go.uber.org/zap/zapcore"
```
Comment (janisz) [inline pkg/features/feature.go:8]: Do we need zap in features flags?

File: pkg/features/features.go [hunk 3]
```diff
@@ -44,3 +50,12 @@ func saveFeature(name, envVar string, defaultValue, overridable bool) FeatureFla
 	Flags[f.envVar] = f
 	return f
 }
+
+// LogFeatureFlags logs the global state of all features flags.
+func LogFeatureFlags() {
+	var context []interface{}
+	for _, flag := range Flags {
```
Comment (janisz) [inline pkg/features/features.go:57]: The order will be random? I think it's better to have it sorted

File: pkg/features/features.go [hunk 4]
```diff
@@ -44,3 +50,12 @@ func saveFeature(name, envVar string, defaultValue, overridable bool) FeatureFla
...
+		context = append(context, flag.LoggingContext())
+	}
+	log.Infow("Feature flags", context...)
```
Comment (janisz) [inline pkg/features/features.go:60]: I think it should be multiline to be easier to read by human and spot configuration on the first glance. How about just `<flag name>: <enabled>` sorted with a flag per line?

File: pkg/features/feature.go [hunk 5]
```diff
@@ -38,3 +40,7 @@ func (f *feature) Enabled() bool {
 		return f.defaultValue
 	}
 }
+
+func (f *feature) LoggingContext() interface{} {
```
Comment (vikin91) [inline pkg/features/feature.go:44]: I am reviewing this and all looks good, except that I cannot find here any context - as known from the `context` pkg. What was the reason for naming it like this?

Comment (stehessel) [reply pkg/features/feature.go:44]: Structured data in logging is also called context. But I can change the name to avoid confusion with golang context.

**Review by vikin91 — APPROVED** [review]
Thanks! I think, you still must make CI happy before merging

**Comment by stehessel** [comment]
Now that https://github.com/stackrox/stackrox/pull/12051 has landed, I rebased and separated the logs into lines for dev-preview and tech-preview as we've discussed.

File: pkg/features/feature.go [hunk 6]
```diff
@@ -3,13 +3,18 @@ package features
...
+	"github.com/stackrox/rox/pkg/logging"
 )
 
 type stage bool
 
 const (
 	devPreview  stage = false
 	techPreview stage = true
+
+	devPreviewString  string = "dev-preview"
```
Comment (janisz) [inline pkg/features/feature.go:16]:
```suggestion
	devPreviewString  = "dev-preview"
```

File: pkg/features/feature.go [hunk 7]
```diff
@@ -49,7 +54,11 @@ func (f *feature) Enabled() bool {
...
 func (f *feature) Stage() string {
 	if f.techPreview {
-		return "tech-preview"
+		return techPreviewString
 	}
-	return "dev-preview"
+	return devPreviewString
+}
+
+func (f *feature) Logging() interface{} {
```
Comment (janisz) [inline pkg/features/feature.go:62]: It's used only in one place it could be unexported or inlined.

**Review by janisz — APPROVED** [review]

**Review by janisz — APPROVED** [review]

**Review by 0x656b694d — COMMENTED** [review]
I believe we could do that nicer by changing the Flags map to a list, and log the flags without sorting. Altering the interface is an unnecessary complication.

**Comment by stehessel** [comment]
For posterity, I inlined the logging function to avoid interface changes. We took a vote re: sorting, and alphabetic sorting come out on top.

**Pull request merged**